scripts/autolabel.lean

```diff
@@ -0,0 +1,150 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jon Eugster, Damiano Testa
+-/
+import Lean.Elab.Command
+
+/-!
+# Automatic labelling of PRs
+
+This file contains the script to automatically assign a GitHub label to a PR.
+
+`AutoLabel.mathlibLabels` contains an assignment of github labels to folders inside
+the mathlib repository. The sctript uses `git diff` to determine which files
+a PR modifies and then finds all labels that should be added based on these changes.
+
+For the time being, the script only adds a label if it finds a single unique label
+that would apply. If multiple labels are found, nothing happens.
+-/
+
+open Lean System
+
+namespace AutoLabel
+
+/--
+A `Label` consists of the
+* The `label` field is the actual GitHub label name.
+* The `dirs` field is the array of all "root paths" such that a modification in a file contained
+  in one of these paths should be labelled with `label`.
+* The `exclusions` field is the array of all "root paths" that are excluded, among the
+  ones that start with the ones in `dirs`.
+  Any modifications to a file in an excluded path is ignored for the purposes of labelling.
+-/
+structure Label where
+  /-- The label name as it appears on github -/
+  label : String
+  /-- Array of paths which fall under this label. e.g. `"Mathlib" / "Algebra"` -/
+  dirs : Array FilePath
+  /-- Array of paths which should be excluded.
+  Any modifications to a file in an excluded path is ignored for the purposes of labelling. -/
+  exclusions : Array FilePath := #[]
+  deriving BEq, Hashable
+
+/--
+Mathlib Labels and their corresponding folders. Add new labels and folders here!
+-/
+def mathlibLabels : Array Label := #[
+  { label := "t-algebra",
+    dirs := #[
+      "Mathlib" / "FieldTheory",
+      "Mathlib" / " RingTheory",
+      "Mathlib" / "GroupTheory",
+      "Mathlib" / "RepresentationTheory",
+      "Mathlib" / "LinearAlgebra"] },
```

Comment:
@grunweg re hiding implementation from the user, this could be a separate JSON file instead of a `def` if you'd prefer that?

Reply:
We could also use an `IO.Ref` and recycle some of the syntax from the other PR, but avoiding Environment extensions.  After all, everything happens in one file only.

Reply:
nvm, that would be less nice with the filepaths...

Reply:
I think the current logic is good enough.